File: src/main/kotlin/codeacademy/bookingforum/app/topicCategory/TopicCategoryMapper.java
package codeacademy.bookingforum.app.topicCategory;

import org.springframework.stereotype.Service;

@Service
public class TopicCategoryMapper {
    public TopicCategoryDto toDto(TopicCategory entity) {
        if (entity == null) {
            return null;
        }
        TopicCategoryDto dto = new TopicCategoryDto();
        dto.setId(entity.getId());
        dto.setTitle(entity.getTitle());
        dto.setDescription(entity.getDescription());
        dto.setRoles(entity.getRoles());
        return dto;
    }

    public TopicCategory fromDto(TopicCategoryDto dto) {
        if (dto == null) {
            return null;
        }
        TopicCategory entity = new TopicCategory();
        entity.setId(dto.getId());
        entity.setTitle(dto.getTitle());
        entity.setDescription(dto.getDescription());
        entity.setRoles(dto.getRoles());
        return entity;
    }
}

File: src/main/kotlin/codeacademy/bookingforum/app/topicCategory/TopicCategoryMapper.kt
package codeacademy.bookingforum.app.topicCategory

import codeacademy.bookingforum.app.topic.Topic
import org.springframework.stereotype.Service

@Service
class TopicCategoryMapper {

    fun fromDtoToEntity(dto: TopicCategoryDto?): TopicCategory? {
        return TopicCategory(
            dto?.id,
            dto?.title,
            dto?.description,
            dto?.role,
            dto?.topics?.map { Topic(it) }
        )
    }

    fun fromEntityToDto(entity: TopicCategory?): TopicCategoryDto? {
        return TopicCategoryDto(
            id = entity?.id,
            title = entity?.title,
            description = entity?.description,
            role = entity?.role,
            topics = entity?.topics?.map { it.id }
        )
    }

    fun fromEntityListToDtoList(entityList: List<TopicCategory?>): List<TopicCategoryDto?> {
        return entityList.map { fromEntityToDto(it) }
    }
}